File: pkg/health/unified_health.go
package health

import (
	"context"
	"encoding/json"
	"time"
)

// HealthStatus 健康状态枚举
type HealthStatus string

const (
	StatusHealthy   HealthStatus = "healthy"   // 健康
	StatusDegraded  HealthStatus = "degraded"  // 降级（部分功能不可用）
	StatusUnhealthy HealthStatus = "unhealthy" // 不健康
)

// UnifiedHealthResponse 统一健康检查响应
type UnifiedHealthResponse struct {
	// 基础字段
	Status    HealthStatus `json:"status"`              // 健康状态
	Service   string       `json:"service"`             // 服务名称
	Version   string       `json:"version"`             // 服务版本
	Timestamp string       `json:"timestamp"`           // 时间戳
	Uptime    int64        `json:"uptime"`              // 运行时间（秒）
	
	// 依赖检查
	Dependencies map[string]DependencyHealth `json:"dependencies,omitempty"` // 依赖服务健康状态
	
	// 系统信息
	System *SystemInfo `json:"system,omitempty"` // 系统信息
	
	// 可选字段
	Message string `json:"message,omitempty"` // 附加消息
}

// DependencyHealth 依赖服务健康状态
type DependencyHealth struct {
	Status    HealthStatus `json:"status"`              // 健康状态
	Latency   int64        `json:"latency,omitempty"`   // 响应延迟（毫秒）
	Message   string       `json:"message,omitempty"`   // 状态消息
	LastCheck string       `json:"last_check,omitempty"` // 最后检查时间
}

// SystemInfo 系统信息
type SystemInfo struct {
	GoVersion     string  `json:"go_version,omitempty"`     // Go版本
	PythonVersion string  `json:"python_version,omitempty"` // Python版本
	OS            string  `json:"os,omitempty"`             // 操作系统
	Arch          string  `json:"arch,omitempty"`           // 架构
	CPUUsage      float64 `json:"cpu_usage,omitempty"`      // CPU使用率（%）
	MemoryUsage   float64 `json:"memory_usage,omitempty"`   // 内存使用率（%）
	Goroutines    int     `json:"goroutines,omitempty"`     // Goroutine数量
}

// ReadinessResponse 就绪检查响应
type ReadinessResponse struct {
	Ready        bool                        `json:"ready"`                  // 是否就绪
	Service      string                      `json:"service"`                // 服务名称
	Timestamp    string                      `json:"timestamp"`              // 时间戳
	Checks       map[string]bool             `json:"checks"`                 // 各项检查结果
	Dependencies map[string]DependencyHealth `json:"dependencies,omitempty"` // 依赖状态
	Message      string                      `json:"message,omitempty"`      // 附加消息
}

// HealthChecker 健康检查器接口
type HealthChecker interface {
	// Check 执行健康检查
	Check(ctx context.Context) *UnifiedHealthResponse
	
	// CheckDependency 检查依赖服务
	CheckDependency(ctx context.Context, name string) *DependencyHealth
	
	// IsReady 检查是否就绪
	IsReady(ctx context.Context) *ReadinessResponse
}

// NewHealthResponse 创建健康响应
func NewHealthResponse(service, version string, startTime time.Time) *UnifiedHealthResponse {
	uptime := time.Since(startTime).Seconds()
	
	return &UnifiedHealthResponse{
		Status:    StatusHealthy,
		Service:   service,
		Version:   version,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Uptime:    int64(uptime),
	}
}

// WithDependencies 添加依赖检查
func (h *UnifiedHealthResponse) WithDependencies(deps map[string]DependencyHealth) *UnifiedHealthResponse {
	h.Dependencies = deps
	
	// 根据依赖状态更新整体状态
	hasUnhealthy := false
	hasDegraded := false
	
	for _, dep := range deps {
		if dep.Status == StatusUnhealthy {
			hasUnhealthy = true
		} else if dep.Status == StatusDegraded {
			hasDegraded = true
		}
	}
	
	if hasUnhealthy {
		h.Status = StatusUnhealthy
	} else if hasDegraded {
		h.Status = StatusDegraded
	}
	
	return h
}

// WithSystem 添加系统信息
func (h *UnifiedHealthResponse) WithSystem(sys *SystemInfo) *UnifiedHealthResponse {
	h.System = sys
	return h
}

// WithMessage 添加消息
func (h *UnifiedHealthResponse) WithMessage(message string) *UnifiedHealthResponse {
	h.Message = message
	return h
}

// ToJSON 转换为JSON
func (h *UnifiedHealthResponse) ToJSON() ([]byte, error) {
	return json.Marshal(h)
}

// NewReadinessResponse 创建就绪响应
func NewReadinessResponse(service string) *ReadinessResponse {
	return &ReadinessResponse{
		Ready:     true,
		Service:   service,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Checks:    make(map[string]bool),
	}
}

// AddCheck 添加检查项
func (r *ReadinessResponse) AddCheck(name string, passed bool) *ReadinessResponse {
	r.Checks[name] = passed
	
	// 只要有一项失败，整体就不就绪
	if !passed {
		r.Ready = false
	}
	
	return r
}

// WithDependencies 添加依赖状态
func (r *ReadinessResponse) WithDependencies(deps map[string]DependencyHealth) *ReadinessResponse {
	r.Dependencies = deps
	
	// 检查依赖是否都健康
	for _, dep := range deps {
		if dep.Status == StatusUnhealthy {
			r.Ready = false
			break
		}
	}
	
	return r
}

// WithMessage 添加消息
func (r *ReadinessResponse) WithMessage(message string) *ReadinessResponse {
	r.Message = message
	return r
}

// ToJSON 转换为JSON
func (r *ReadinessResponse) ToJSON() ([]byte, error) {
	return json.Marshal(r)
}

// CheckDependencyWithTimeout 带超时的依赖检查
func CheckDependencyWithTimeout(ctx context.Context, name string, checkFn func() error, timeout time.Duration) *DependencyHealth {
	start := time.Now()
	
	// 创建带超时的context
	checkCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	
	// 执行检查
	errCh := make(chan error, 1)
	go func() {
		errCh <- checkFn()
	}()
	
	var err error
	select {
	case err = <-errCh:
		// 检查完成
	case <-checkCtx.Done():
		// 超时
		err = checkCtx.Err()
	}
	
	latency := time.Since(start).Milliseconds()
	
	// 构建响应
	health := &DependencyHealth{
		Latency:   latency,
		LastCheck: time.Now().UTC().Format(time.RFC3339),
	}
	
	if err != nil {
		health.Status = StatusUnhealthy
		health.Message = err.Error()
	} else {
		health.Status = StatusHealthy
		health.Message = "OK"
	}
	
	return health
}

// StandardHealthChecks 标准健康检查项
var StandardHealthChecks = struct {
	Database      string
	Redis         string
	MessageQueue  string
	ExternalAPI   string
	FileStorage   string
	VectorDB      string
	GraphDB       string
	SearchEngine  string
}{
	Database:     "database",
	Redis:        "redis",
	MessageQueue: "message_queue",
	ExternalAPI:  "external_api",
	FileStorage:  "file_storage",
	VectorDB:     "vector_db",
	GraphDB:      "graph_db",
	SearchEngine: "search_engine",
}

File: algo/common/unified_health.py
"""
统一健康检查格式 - Python版本

与Go服务保持一致的健康检查格式
"""

import asyncio
import platform
import sys
import time
from datetime import datetime
from enum import Enum
from typing import Callable, Dict, Optional

import psutil
from pydantic import BaseModel, Field


class HealthStatus(str, Enum):
    """健康状态枚举"""

    HEALTHY = "healthy"
    DEGRADED = "degraded"
    UNHEALTHY = "unhealthy"


class DependencyHealth(BaseModel):
    """依赖服务健康状态"""

    status: HealthStatus = Field(..., description="健康状态")
    latency: Optional[int] = Field(None, description="响应延迟（毫秒）")
    message: Optional[str] = Field(None, description="状态消息")
    last_check: Optional[str] = Field(None, description="最后检查时间")

    class Config:
        json_schema_extra = {
            "example": {
                "status": "healthy",
                "latency": 15,
                "message": "OK",
                "last_check": "2025-11-01T10:30:00Z",
            }
        }


class SystemInfo(BaseModel):
    """系统信息"""

    python_version: Optional[str] = Field(None, description="Python版本")
    os: Optional[str] = Field(None, description="操作系统")
    arch: Optional[str] = Field(None, description="架构")
    cpu_usage: Optional[float] = Field(None, description="CPU使用率（%）")
    memory_usage: Optional[float] = Field(None, description="内存使用率（%）")
    process_memory: Optional[float] = Field(None, description="进程内存（MB）")

    class Config:
        json_schema_extra = {
            "example": {
                "python_version": "3.11.5",
                "os": "Linux",
                "arch": "x86_64",
                "cpu_usage": 45.2,
                "memory_usage": 62.8,
                "process_memory": 512.5,
            }
        }


class UnifiedHealthResponse(BaseModel):
    """统一健康检查响应"""

    status: HealthStatus = Field(..., description="健康状态")
    service: str = Field(..., description="服务名称")
    version: str = Field(..., description="服务版本")
    timestamp: str = Field(
        default_factory=lambda: datetime.utcnow().isoformat() + "Z", description="时间戳"
    )
    uptime: int = Field(..., description="运行时间（秒）")

    # 可选字段
    dependencies: Optional[Dict[str, DependencyHealth]] = Field(None, description="依赖服务健康状态")
    system: Optional[SystemInfo] = Field(None, description="系统信息")
    message: Optional[str] = Field(None, description="附加消息")

    class Config:
        json_schema_extra = {
            "example": {
                "status": "healthy",
                "service": "agent-engine",
                "version": "1.0.0",
                "timestamp": "2025-11-01T10:30:00Z",
                "uptime": 3600,
                "dependencies": {
                    "redis": {
                        "status": "healthy",
                        "latency": 2,
                        "message": "OK",
                    }
                },
            }
        }


class ReadinessResponse(BaseModel):
    """就绪检查响应"""

    ready: bool = Field(..., description="是否就绪")
    service: str = Field(..., description="服务名称")
    timestamp: str = Field(
        default_factory=lambda: datetime.utcnow().isoformat() + "Z", description="时间戳"
    )
    checks: Dict[str, bool] = Field(default_factory=dict, description="各项检查结果")
    dependencies: Optional[Dict[str, DependencyHealth]] = Field(None, description="依赖状态")
    message: Optional[str] = Field(None, description="附加消息")

    class Config:
        json_schema_extra = {
            "example": {
                "ready": True,
                "service": "agent-engine",
                "timestamp": "2025-11-01T10:30:00Z",
                "checks": {
                    "database": True,
                    "redis": True,
                    "llm": True,
                },
            }
        }


class HealthChecker:
    """健康检查器"""

    def __init__(self, service_name: str, version: str, start_time: float | None = None):
        """
        初始化健康检查器

        Args:
            service_name: 服务名称
            version: 服务版本
            start_time: 启动时间（时间戳）
        """
        self.service_name = service_name
        self.version = version
        self.start_time = start_time or time.time()

    def get_uptime(self) -> int:
        """获取运行时间（秒）"""
        return int(time.time() - self.start_time)

    def get_system_info(self) -> SystemInfo:
        """获取系统信息"""
        try:
            cpu_percent = psutil.cpu_percent(interval=0.1)
            memory = psutil.virtual_memory()
            process = psutil.Process()
            process_memory_mb = process.memory_info().rss / 1024 / 1024

            return SystemInfo(
                python_version=f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
                os=platform.system(),
                arch=platform.machine(),
                cpu_usage=round(cpu_percent, 2),
                memory_usage=round(memory.percent, 2),
                process_memory=round(process_memory_mb, 2),
            )
        except Exception:
            return SystemInfo(
                python_version=f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
                os=platform.system(),
                arch=platform.machine(),
            )

    async def check_dependency(
        self,
        name: str,
        check_fn: Callable[[], bool] | Callable[[], asyncio.Future],
        timeout: float = 5.0,
    ) -> DependencyHealth:
        """
        检查依赖服务

        Args:
            name: 依赖服务名称
            check_fn: 检查函数
            timeout: 超时时间（秒）

        Returns:
            依赖健康状态
        """
        start_time = time.time()

        try:
            # 检查函数是否是协程
            import inspect

            if inspect.iscoroutinefunction(check_fn):
                result = await asyncio.wait_for(check_fn(), timeout=timeout)
            else:
                result = check_fn()

            latency = int((time.time() - start_time) * 1000)

            return DependencyHealth(
                status=HealthStatus.HEALTHY if result else HealthStatus.UNHEALTHY,
                latency=latency,
                message="OK" if result else "Check failed",
                last_check=datetime.utcnow().isoformat() + "Z",
            )

        except asyncio.TimeoutError:
            latency = int(timeout * 1000)
            return DependencyHealth(
                status=HealthStatus.UNHEALTHY,
                latency=latency,
                message="Timeout",
                last_check=datetime.utcnow().isoformat() + "Z",
            )

        except Exception as e:
            latency = int((time.time() - start_time) * 1000)
            return DependencyHealth(
                status=HealthStatus.UNHEALTHY,
                latency=latency,
                message=str(e),
                last_check=datetime.utcnow().isoformat() + "Z",
            )

    async def check(
        self,
        dependencies: Dict[str, Callable] | None = None,
        include_system: bool = True,
    ) -> UnifiedHealthResponse:
        """
        执行健康检查

        Args:
            dependencies: 依赖检查函数字典
            include_system: 是否包含系统信息

        Returns:
            健康检查响应
        """
        # 检查依赖
        dep_health = {}
        if dependencies:
            for name, check_fn in dependencies.items():
                dep_health[name] = await self.check_dependency(name, check_fn)

        # 确定整体状态
        status = HealthStatus.HEALTHY
        if dep_health:
            has_unhealthy = any(d.status == HealthStatus.UNHEALTHY for d in dep_health.values())
            has_degraded = any(d.status == HealthStatus.DEGRADED for d in dep_health.values())

            if has_unhealthy:
                status = HealthStatus.UNHEALTHY
            elif has_degraded:
                status = HealthStatus.DEGRADED

        # 构建响应
        response = UnifiedHealthResponse(
            status=status,
            service=self.service_name,
            version=self.version,
            uptime=self.get_uptime(),
            dependencies=dep_health if dep_health else None,
            system=self.get_system_info() if include_system else None,
        )

        return response

    async def is_ready(
        self,
        checks: Dict[str, Callable] | None = None,
        dependencies: Dict[str, Callable] | None = None,
    ) -> ReadinessResponse:
        """
        检查是否就绪

        Args:
            checks: 检查项函数字典
            dependencies: 依赖检查函数字典

        Returns:
            就绪检查响应
        """
        # 执行各项检查
        check_results = {}
        if checks:
            for name, check_fn in checks.items():
                try:
                    import inspect

                    if inspect.iscoroutinefunction(check_fn):
                        result = await check_fn()
                    else:
                        result = check_fn()
                    check_results[name] = bool(result)
                except Exception:
                    check_results[name] = False

        # 检查依赖
        dep_health = {}
        if dependencies:
            for name, check_fn in dependencies.items():
                dep_health[name] = await self.check_dependency(name, check_fn)

        # 确定是否就绪
        ready = all(check_results.values()) if check_results else True

        # 如果有不健康的依赖，也标记为不就绪
        if dep_health and any(d.status == HealthStatus.UNHEALTHY for d in dep_health.values()):
            ready = False

        return ReadinessResponse(
            ready=ready,
            service=self.service_name,
            checks=check_results,
            dependencies=dep_health if dep_health else None,
        )


# 标准健康检查项名称
class StandardHealthChecks:
    """标准健康检查项"""

    DATABASE = "database"
    REDIS = "redis"
    MESSAGE_QUEUE = "message_queue"
    EXTERNAL_API = "external_api"
    FILE_STORAGE = "file_storage"
    VECTOR_DB = "vector_db"
    GRAPH_DB = "graph_db"
    SEARCH_ENGINE = "search_engine"
    LLM = "llm"
    EMBEDDING = "embedding"


# FastAPI集成示例
"""
from fastapi import FastAPI
from unified_health import HealthChecker, StandardHealthChecks

app = FastAPI()
health_checker = HealthChecker("my-service", "1.0.0")

@app.get("/health")
async def health():
    '''健康检查'''
    return await health_checker.check(
        dependencies={
            StandardHealthChecks.REDIS: lambda: redis_client.ping(),
            StandardHealthChecks.DATABASE: lambda: db.is_connected(),
        }
    )

@app.get("/ready")
async def readiness():
    '''就绪检查'''
    return await health_checker.is_ready(
        checks={
            "llm": lambda: llm_client is not None,
            "tools": lambda: tool_registry is not None,
        },
        dependencies={
            StandardHealthChecks.REDIS: lambda: redis_client.ping(),
        }
    )
"""

